File: src/main/java/com/demonwav/mcdev/insight/generation/ui/EventListenerWizard.java
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2017 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.insight.generation.ui;

import com.intellij.ide.highlighter.JavaHighlightingColors;
import com.intellij.openapi.editor.ex.util.EditorUtil;
import com.intellij.openapi.wm.ex.IdeFocusTraversalPolicy;
import com.intellij.uiDesigner.core.GridConstraints;
import com.intellij.util.ui.UIUtil;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JSeparator;
import javax.swing.JTextField;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class EventListenerWizard {
    private JPanel panel;
    private JTextField classNameTextField;
    // TODO: This field isn't always focusable when this wizard is created. I have no idea why or how to fix it, but it needs to be fixed before release
    private JTextField listenerNameTextField;
    private JLabel publicVoidLabel;
    private JPanel contentPanel;
    private JSeparator separator;

    private static final GridConstraints innerContentPanelConstraints = new GridConstraints();
    static {
        innerContentPanelConstraints.setRow(0);
        innerContentPanelConstraints.setColumn(0);
        innerContentPanelConstraints.setRowSpan(1);
        innerContentPanelConstraints.setColSpan(1);
        innerContentPanelConstraints.setAnchor(GridConstraints.ANCHOR_CENTER);
        innerContentPanelConstraints.setFill(GridConstraints.FILL_BOTH);
        innerContentPanelConstraints.setHSizePolicy(GridConstraints.SIZEPOLICY_FIXED);
        innerContentPanelConstraints.setVSizePolicy(GridConstraints.SIZEPOLICY_FIXED);
    }

    public EventListenerWizard(@Nullable JPanel panel, @NotNull String className, @NotNull String defaultListenerName) {
        classNameTextField.setFont(EditorUtil.getEditorFont());
        listenerNameTextField.setFont(EditorUtil.getEditorFont());
        publicVoidLabel.setFont(EditorUtil.getEditorFont());
        if (UIUtil.isUnderDarcula()) {
            publicVoidLabel.setForeground(JavaHighlightingColors.KEYWORD.getDefaultAttributes().getForegroundColor());
        } else {
            publicVoidLabel.setForeground(JavaHighlightingColors.KEYWORD.getFallbackAttributeKey().getDefaultAttributes().getForegroundColor());
        }

        if (panel != null) {
            separator.setVisible(true);
            contentPanel.add(panel, innerContentPanelConstraints);
        }

        classNameTextField.setText(className);
        listenerNameTextField.setText(defaultListenerName);

        IdeFocusTraversalPolicy.getPreferredFocusedComponent(listenerNameTextField).requestFocus();
        listenerNameTextField.requestFocus();
    }

    public JPanel getPanel() {
        return panel;
    }

    public String getChosenClassName() {
        return listenerNameTextField.getText();
    }
}

File: src/main/kotlin/com/demonwav/mcdev/insight/generation/ui/EventListenerWizard.kt
/*
 * Minecraft Dev for IntelliJ
 *
 * https://minecraftdev.org
 *
 * Copyright (c) 2019 minecraft-dev
 *
 * MIT License
 */

package com.demonwav.mcdev.insight.generation.ui

import com.intellij.ide.highlighter.JavaHighlightingColors
import com.intellij.openapi.editor.ex.util.EditorUtil
import com.intellij.openapi.wm.ex.IdeFocusTraversalPolicy
import com.intellij.uiDesigner.core.GridConstraints
import com.intellij.util.ui.UIUtil
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JSeparator
import javax.swing.JTextField

class EventListenerWizard(panel: JPanel?, className: String, defaultListenerName: String) {
    lateinit var panel: JPanel
    private lateinit var classNameTextField: JTextField
    private lateinit var listenerNameTextField: JTextField
    private lateinit var publicVoidLabel: JLabel
    private lateinit var contentPanel: JPanel
    private lateinit var separator: JSeparator

    init {
        classNameTextField.font = EditorUtil.getEditorFont()
        listenerNameTextField.font = EditorUtil.getEditorFont()
        publicVoidLabel.font = EditorUtil.getEditorFont()
        if (UIUtil.isUnderDarcula()) {
            publicVoidLabel.foreground = JavaHighlightingColors.KEYWORD.defaultAttributes.foregroundColor
        } else {
            publicVoidLabel.foreground =
                JavaHighlightingColors.KEYWORD.fallbackAttributeKey!!.defaultAttributes.foregroundColor
        }

        if (panel != null) {
            separator.isVisible = true
            contentPanel.add(panel, innerContentPanelConstraints)
        }

        classNameTextField.text = className
        listenerNameTextField.text = defaultListenerName

        IdeFocusTraversalPolicy.getPreferredFocusedComponent(listenerNameTextField).requestFocus()
        listenerNameTextField.requestFocus()
    }

    val chosenClassName: String
        get() = listenerNameTextField.text

    companion object {
        private val innerContentPanelConstraints = GridConstraints()

        init {
            innerContentPanelConstraints.row = 0
            innerContentPanelConstraints.column = 0
            innerContentPanelConstraints.rowSpan = 1
            innerContentPanelConstraints.colSpan = 1
            innerContentPanelConstraints.anchor = GridConstraints.ANCHOR_CENTER
            innerContentPanelConstraints.fill = GridConstraints.FILL_BOTH
            innerContentPanelConstraints.hSizePolicy = GridConstraints.SIZEPOLICY_FIXED
            innerContentPanelConstraints.vSizePolicy = GridConstraints.SIZEPOLICY_FIXED
        }
    }
}
